htlcswitch/circuit.go

```diff
@@ -1,135 +1,170 @@
+	hashIndex map[[32]byte][]*PaymentCircuit
 }
 
-// newCircuitMap creates a new instance of the circuitMap.
-func newCircuitMap() *circuitMap {
-	return &circuitMap{
-		circuits: make(map[circuitKey]*paymentCircuit),
+// NewCircuitMap creates a new instance of the CircuitMap.
+func NewCircuitMap() *CircuitMap {
+	return &CircuitMap{
+		circuits:  make(map[circuitKey]*PaymentCircuit),
+		hashIndex: make(map[[32]byte][]*PaymentCircuit),
 	}
 }
 
-// add adds a new active payment circuit to the circuitMap.
-func (m *circuitMap) add(circuit *paymentCircuit) error {
-	m.Lock()
-	defer m.Unlock()
-
-	// Examine the circuit map to see if this circuit is already in use or
-	// not. If so, then we'll simply increment the reference count.
-	// Otherwise, we'll create a new circuit from scratch.
-	//
-	// TODO(roasbeef): include dest+src+amt in key
-	if c, ok := m.circuits[circuit.PaymentHash]; ok {
-		c.RefCount++
-		return nil
+// LookupByHTLC looks up the payment circuit by the outgoing channel and HTLC
+// IDs. Returns nil if there is no such circuit.
+func (cm *CircuitMap) LookupByHTLC(chanID lnwire.ShortChannelID, htlcID uint64) *PaymentCircuit {
+	cm.mtx.RLock()
+
+	key := circuitKey{
+		chanID: chanID,
+		htlcID: htlcID,
 	}
+	circuit := cm.circuits[key]
+
+	cm.mtx.RUnlock()
+	return circuit
+}
+
+// LookupByPaymentHash looks up and returns any payment circuits with a given
+// payment hash.
+func (cm *CircuitMap) LookupByPaymentHash(hash [32]byte) []*PaymentCircuit {
+	cm.mtx.RLock()
+	circuits := cm.hashIndex[hash]
+	cm.mtx.RUnlock()
+	return circuits
+}
+
+// Add adds a new active payment circuit to the CircuitMap.
+func (cm *CircuitMap) Add(circuit *PaymentCircuit) error {
+	cm.mtx.Lock()
 
-	m.circuits[circuit.PaymentHash] = circuit
+	key := circuitKey{
+		chanID: circuit.OutgoingChanID,
+		htlcID: circuit.OutgoingHTLCID,
+	}
+	cm.circuits[key] = circuit
+	cm.hashIndex[circuit.PaymentHash] =
+		append(cm.hashIndex[circuit.PaymentHash], circuit)
 
+	cm.mtx.Unlock()
 	return nil
 }
 
-// remove destroys the target circuit by removing it from the circuit map.
-func (m *circuitMap) remove(key circuitKey) (*paymentCircuit, error) {
-	m.Lock()
-	defer m.Unlock()
+// Remove destroys the target circuit by removing it from the circuit map.
+func (cm *CircuitMap) Remove(chanID lnwire.ShortChannelID, htlcID uint64) error {
+	cm.mtx.Lock()
+	defer cm.mtx.Unlock()
+
+	// Look up circuit so that pointer can be matched in the hash index.
+	key := circuitKey{
+		chanID: chanID,
+		htlcID: htlcID,
+	}
+	circuit, found := cm.circuits[key]
+	if !found {
+		return errors.Errorf("Can't find circuit for HTLC %v", key)
+	}
+
+	delete(cm.circuits, key)
 
-	if circuit, ok := m.circuits[key]; ok {
-		if circuit.RefCount--; circuit.RefCount == 0 {
-			delete(m.circuits, key)
+	// Remove key from hash index.
+	index := -1
+	circuitsWithHash := cm.hashIndex[circuit.PaymentHash]
+	for i, indexedCircuit := range circuitsWithHash {
+		if indexedCircuit == circuit {
```

Comment:
I'll do it if you really want, but I'm not convinced it'll be better for this use case. Firstly, LookupByPaymentHash has to return a slice. Second, I expect most of the time payment hashes will not be reused and there will be one circuit per payment hash, or at least a very small number.

Reply:
Well, it returns a slice currently but doesn't have to necessarily. In fact, I think a slice should be derived from the map during every call to `LookupByPaymentHash`. I just noticed that we return the slices directly, whose contents be mutated by subsequent calls to `Remove`, which does an in-place swap/truncate. Using a map will also ensure that `Add` and `Remove` are idempotent, this implementation would break if they are not called the same number of times.

Reply:
I think as is you could cause a nil pointer deference by
```
circuits := cm.LookupByPaymentHash(hash)
cm.Remove(circuits[0])
for i, circuit := range circuits {
    // Will fail at i = len(circuits) - 1
    circuit.variable
}
```